### ISSUE TITLE
Improve the reliability of SyncContext.PullItemsAsync

### ISSUE THREAD

**Issue by asierpn** [Azure/azure-mobile-apps#878]
**Is your feature request related to a problem? Please describe.**

I was dealing with the performance of PullItemsAsync with large datasets and SQLite until I found the WriteDeltaTokenInterval, I have set it to 2000, so now pulling a table with 4500 items takes less than 1 second, whereas before this change it took >40 seconds, but I am concerned about the possibility of getting a corrupted database.

**Describe the solution you'd like**

I think that the possibility of corruption still exists if you process the items one at a time, although it is minimal, and the way to avoid it completely is to combine the DeleteAsync, UpsertAsync and SetDeltaTokenAsync methods and perform all these commands in the same transaction in the OfflineStore

**Comment by adrianhall** [comment]
You’d think, but experimentation says that transactions actually slow down the process.  I’m not saying it’s logical (it isn’t), but reality is that I’ve coded up what you suggest as the solution, and it slows the process down in large datasets.

**Comment by asierpn** [comment]
Yes, it is strange, because the Upsert is done in a transaction an it works fine with 2000 item batches, the change is that the item deletions and the DeltaToken insert/update are done within the same transaction.

**Comment by adrianhall** [comment]
So, I was delving into this...

The code in the SyncContext does this:

```csharp
updatedAt = ServiceSerializer.GetUpdatedAt(item)?.ToUniversalTime();
if (itemCount % options.WriteDeltaTokenInterval == 0)
{
  await OfflineStore.DeleteAsync(tableName, deletedIds);
  deletedIds.ForEach(x => SendItemWasStoredEvent(tableName, x, itemCount, expectedItems));
  deletedIds.Clear();

  await OfflineStore.UpsertAsync(tableName, upsertList.Values, ignoreMissingColumns: true);
  upsertList.Keys.ToList().ForEach(x => SendItemWasStoredEvent(tableName, x, itemCount, expectedItems));
  upsertList.Clear();

  if (updatedAt.HasValue)
  {
    await DeltaTokenStore.SetDeltaTokenAsync(tableName, queryId, updatedAt.Value);
    updatedAt = null;
  }
}
```

See https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.Datasync.Client/Offline/SyncContext.cs#L395

All this to say that all the deleted items are bunched up (in the same way as the Upsert) and sent to the SqliteStore together based on how often you write the delta token out (which is the batch size).  So, no difference there.  Now - if we look at the code in the SqliteStore (see https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.Datasync.Client/Offline/SyncContext.cs#L395):

```csharp
string sql = SqlStatements.DeleteFromTable(query, out Dictionary<string, object> parameters);
using (operationLock.AcquireLock())
{
  ExecuteNonQueryInternal(sql, parameters);
}
```

Which is a long way of saying that we delete the IDs using an operation lock AND a single statement, so it's as fast as we can make it anyhow.  We don't need a transaction because it's not needed because we only use one SQL statement anyhow.

There is always the possibility of a corrupted database, but that's true irrespective of the use of transactions.  If you kill the app (or the app crashes) while writing the database file, there isn't a lot we can do about it.

Long story short - nothing we can do here.

**Comment by asierpn** [comment]
Thanks for the detailed explanation, I thought the possibility of a corrupted database came from the scenario where the the app is closed between the upserts and the delta token update.
